splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java

```diff
@@ -69,211 +137,296 @@
-                                                this.currentColumn, quoteScopeStartingLine,getLineNumber()));
-                    }
-
-                    currentRow.append(line); // update untokenized CSV row
-
-                    if(line.isEmpty()){
-                        // consecutive newlines
-                        continue;
-                    }
-                }
+        runStateMachine(columns, quotedColumns);
+        return true;
+    }
+
+    List<Integer> getExactColumnSizes() {
+        return exactColumnSizes;
+    }
+
+    private void runStateMachine(final List<String> columns, final BooleanList quotedColumns) throws IOException {
+        while (true) {
+            switch (state) {
+                case NORMAL:
+                    handleNormalState(columns, quotedColumns);
+                    break;
+                case QUOTE_MODE:
+                    handleQuoteState();
+                    break;
+                case FINISHED:
+                    return;
+            }
+        }
+    }
+
+    private void appendToRowNewLine() {
+        if (checkExactSizeFirst) {
+            currentRow.append(String.valueOf(NEWLINE));
+        } else if (exactColumnSizes != null) {
+            rowIdx++; // simulate adding a newline by moving a sequence by one
+        } else {
+            currentRow.append(String.valueOf(NEWLINE));
+        }
+    }
+
+    private void appendToRowFromFile() throws IOException {
+        if (checkExactSizeFirst) {
+            currentRow.append(readLine());
+        } else if (exactColumnSizes != null) {
+            rowIdx++; // simulate adding a newline by moving a sequence by one
+        } else {
+            currentRow.append(readLine());
+        }
+    }
+
+    private void addToColumn(char c) {
+        if (checkExactSizeFirst) {
+            if (colIdx >= exactColumnSizes.size()) {
+                // it seems this could happen, still we want to tolerate it and not throw. (see expected behavior in HdfsImportIT::testNullDatesWithMixedCaseAccuracy)
+                return;
             }
+            exactColumnSizes.set(colIdx, exactColumnSizes.get(colIdx) + 1);
+        } else {
+            currentColumn.append(c);
+        }
+    }
 
-            final char c=line.charAt(charIndex);
-
-            if(TokenizerState.NORMAL.equals(state)){
-                if(c==delimeterChar){
-					/*
-					 * Delimiter. Save the column (trim trailing space if required) then continue to next character.
-					 */
-                    if(!surroundingSpacesNeedQuotes){
-                        appendSpaces(currentColumn,potentialSpaces);
-                    }
-                    columns.add(currentColumn.length()>0?currentColumn.toString():null); // "" -> null
-                    potentialSpaces=0;
+    private void allocateColumnMemory() {
+        if (!checkExactSizeFirst) {
```

Comment:
I think we reuse the `currentColumn` StringBuilder for every column, right? We could initialize the StringBuilder to the maximum size here and keep reusing it (by calling `.setLength(0)`), we don't have to call `allocateColumnMemory` for every column, right?

Reply:
Good point, I actually thought about that but I thought at the beginning that wiping out `StringBuilder` to reuse it could be as expensive as creating a new one. Now that I read more about it, I think it makes sense to reuse if possible. Thanks!